src/machi_file_proxy.erl

```diff
@@ -664,3 +659,26 @@ do_write(FHd, CsumTable, Filename, TaggedCsum, Offset, Size, Data) ->
             [Other, Filename, Offset, Size]),
             {error, Other}
     end.
+
+%% @doc Trim both right and left border of chunks to fit in to given
+%% range [LeftPos, RightPos]. TODO: write unit tests for this function.
+slice_both_side([], _, _) -> [];
+slice_both_side([{F, Offset, Chunk, _Csum}|L], LeftPos, RightPos)
+  when Offset < LeftPos andalso LeftPos < RightPos ->
+    TrashLen = 8 * (LeftPos - Offset),
+    <<_:TrashLen/binary, NewChunk/binary>> = Chunk,
+    NewChecksum = machi_util:make_tagged_csum(client_sha, Chunk),
+    NewH = {F, LeftPos, NewChunk, NewChecksum},
+    slice_both_side([NewH|L], LeftPos, RightPos);
+slice_both_side(Chunks, LeftPos, RightPos) when LeftPos =< RightPos ->
+    %% TODO: optimize
+    [{F, Offset, Chunk, _Csum}|L] = lists:reverse(Chunks),
+    Size = iolist_size(Chunk),
+    if RightPos < Offset + Size ->
+            NewSize = RightPos - Offset,
+            <<NewChunk:NewSize/binary, _/binary>> = Chunk,
+            NewChecksum = machi_util:make_tagged_csum(client_sha, Chunk),
```

Comment:
Hrm, perhaps ?CSUM_TAG_SERVER_REGEN_SHA.  To be honest, I don't have 100% solid definitions for each of the checksum types in `include/machi.hrl.  But we probably should create a good definition for each one!

When I first wrote them, my feeling was that ?CSUM_TAG_SERVER_SHA would be created by the server when the client was stupid and gave the server a ?CSUM_TAG_NONE.  Perhaps the data was corrupted by the network or some other place, silly client!  So ?CSUM_TAG_SERVER_SHA is a substitute for an original, client-generated checksum.

Then ?CSUM_TAG_SERVER_REGEN_SHA would be used whenever the server had verified an original client_sha or server_sha but had to recalculate due to a trim or a read that is not at an original chunk offset + size.